packages/cli/src/audit/risks/credentials.risk.ts

```diff
@@ -0,0 +1,133 @@
+import { DateUtils } from 'typeorm/util/DateUtils';
+import * as Db from '@/Db';
+import config from '@/config';
+import { CREDENTIALS_REPORT } from '@/audit/constants';
+import type { WorkflowEntity } from '@/databases/entities/WorkflowEntity';
+import type { Risk } from '@/audit/types';
+
+async function getAllCredsInUse(workflows: WorkflowEntity[]) {
+	const credsInAnyUse = new Set<string>();
+	const credsInActiveUse = new Set<string>();
+
+	workflows.forEach((workflow) => {
+		workflow.nodes.forEach((node) => {
+			if (!node.credentials) return;
+
+			Object.values(node.credentials).forEach((cred) => {
+				if (!cred?.id) return;
+
+				credsInAnyUse.add(cred.id.toString());
+
+				if (workflow.active) credsInActiveUse.add(cred.id.toString());
+			});
+		});
+	});
+
+	return {
+		credsInAnyUse,
+		credsInActiveUse,
+	};
+}
+
+async function getAllExistingCreds() {
+	return Db.collections.Credentials.find().then((c) =>
+		c.map((i) => ({ kind: 'credential' as const, id: i.id.toString(), name: i.name })),
+	);
+}
+
+async function getExecutionsInPastDays(days: number) {
+	const date = new Date();
+
+	date.setDate(date.getDate() - days);
+
+	const utcDate = DateUtils.mixedDateToUtcDatetimeString(date) as string;
+
+	return Db.collections.Execution.find({
+		select: ['workflowData'],
+		where: {
+			startedAt: MoreThanOrEqual(utcDate),
+		},
+	});
+}
+
+/**
+ * Return IDs of credentials in worfklows executed in the past n days.
+ */
+async function getCredsInRecentlyExecutedWorkflows(days: number) {
+	const recentExecutions = await getExecutionsInPastDays(days);
+
+	return recentExecutions.reduce<Set<string>>((acc, execution) => {
+		execution.workflowData?.nodes.forEach((node) => {
+			if (node.credentials) {
+				Object.values(node.credentials).forEach((c) => {
+					if (c.id) acc.add(c.id.toString());
+				});
+			}
+		});
+
+		return acc;
+	}, new Set());
+}
+
+export async function reportCredentialsRisk(workflows: WorkflowEntity[]) {
+	const days = config.getEnv('security.audit.daysAbandonedWorkflow');
+
+	const allExistingCreds = await getAllExistingCreds();
+	const { credsInAnyUse, credsInActiveUse } = await getAllCredsInUse(workflows);
+	const recentlyExecutedCreds = await getCredsInRecentlyExecutedWorkflows(days);
+
+	const credsNotInAnyUse = allExistingCreds.filter((c) => !credsInAnyUse.has(c.id));
+	const credsNotInActiveUse = allExistingCreds.filter((c) => !credsInActiveUse.has(c.id));
+	const credsNotRecentlyExecuted = allExistingCreds.filter((c) => !recentlyExecutedCreds.has(c.id));
+
+	const issues = [credsNotInAnyUse, credsNotInActiveUse, credsNotRecentlyExecuted];
+
+	if (issues.every((i) => i.length === 0)) return null;
+
+	const report: Risk.StandardReport = {
+		risk: CREDENTIALS_REPORT.RISK,
+		sections: [],
+	};
+
+	const hint = 'Keeping unused credentials in your instance is an unneeded security risk.';
+	const recommendation = 'Consider deleting these credentials if you no longer need them.';
```

Comment:
So I began with this but started having second thoughts. A large number of report messages are constructed based on conditions, so we end up with some messages as constants and some messages inside the logic (or moving also the logic away, not very clean) creating some indirection and a lot of inconsistency.

Since we do not offer i18n on the BE, the upside looks a bit limited, so I'd prefer to keep them in the reports, but let me know otherwise and I'll move them.

Reply:
For reference, titles and sections are in constants only so they can be used by tests.